packages/client/src/runtime/core/engines/common/resolveEnginePath.ts

```diff
@@ -0,0 +1,125 @@
+import Debug from '@prisma/debug'
+import { getEnginesPath } from '@prisma/engines'
+import { getNodeAPIName, getPlatform, Platform } from '@prisma/get-platform'
+import { plusX } from '@prisma/internals'
+import fs from 'fs'
+import path from 'path'
+
+import { PrismaClientInitializationError } from '../../errors/PrismaClientInitializationError'
+import { EngineConfig } from './Engine'
+import { binaryTargetsWasIncorrectlyPinned } from './errors/engine-not-found/binaryTargetsWasIncorrectlyPinned'
+import { bundlerHasTamperedWithEngineCopy } from './errors/engine-not-found/bundlerHasTamperedWithEngineCopy'
+import { EngineNotFoundErrorInput } from './errors/engine-not-found/EngineNotFoundErrorInput'
+import { nativeGeneratedOnDifferentPlatform } from './errors/engine-not-found/nativeGeneratedOnDifferentPlatform'
+import { toolingHasTamperedWithEngineCopy } from './errors/engine-not-found/toolingHasTamperedWithEngineCopy'
+
+const debug = Debug('prisma:client:engines:resolveEnginePath')
+
+/**
+ * Resolves the path of a given engine type (binary or library) and config. If
+ * the engine could not be found, we will try to help the user by providing
+ * helpful error messages.
+ * @param engineType
+ * @param config
+ * @returns
+ */
+export async function resolveEnginePath(engineType: 'binary' | 'library', config: EngineConfig) {
+  // if the user provided a custom path, or if engine previously found
+  const prismaPath = {
+    binary: process.env.PRISMA_QUERY_ENGINE_BINARY ?? config.prismaPath,
+    library: process.env.PRISMA_QUERY_ENGINE_LIBRARY ?? config.prismaPath,
+  }[engineType]
+
+  if (prismaPath !== undefined) return prismaPath
+
+  // otherwise we will search to find the nearest query engine file
+  const { enginePath, searchedLocations } = await findEnginePath(engineType, config)
+
+  debug('enginePath', enginePath)
+
+  // if we find it, we apply +x chmod to the binary, cache, and return
+  if (enginePath !== undefined && engineType === 'binary') plusX(enginePath)
+  if (enginePath !== undefined) return (config.prismaPath = enginePath)
+
+  // if we don't find it, then we will throw helpful error messages
+  const binaryTarget = await getPlatform()
+  const generatorBinaryTargets = config.generator?.binaryTargets ?? []
+  const hasNativeBinaryTarget = generatorBinaryTargets.some((bt) => bt.native === true)
+  const hasMissingBinaryTarget = !generatorBinaryTargets.some((bt) => bt.value === binaryTarget)
+  const clientHasBeenBundled = __filename.match(/library|binary\.m?js/) === null // runtime name
```

Comment:
I wonder if we can make this more future proof - if we'd add new runtime files or change the names of existing ones it is quite easy to forget about this line existence. And since the only thing it affects is the error message, it could be quite a while before we notice it.

How about, for example, setting the name of the current runtime in a global during build using `define` and changing this check to something like `__filename.includes(CURRENT_RUNTIME_NAME)` ? That way we'd at least see it the build config and will remember to change it if we change output name.

Reply:
The tests should currently catch that, but I think it would be cleaner that way. Will add that to the changes.